gadopt/equations.py

```diff
@@ -45,14 +45,21 @@ def __init__(self, test_space, trial_space, quad_degree=None, **kwargs):
 
         self.dx = firedrake.dx(domain=self.mesh, degree=quad_degree)
 
+        self.kwargs = kwargs
+
         # self._terms stores the actual instances of the BaseTerm-classes in self.terms
         self._terms = []
         for TermClass in self.terms:
             self._terms.append(TermClass(test_space, trial_space, self.dx, self.ds, self.dS, **kwargs))
 
     def mass_term(self, test, trial):
         r"""Return the UFL for the mass term \int test * trial * dx typically used in the time term."""
-        return firedrake.inner(test, trial) * self.dx
+
+        if 'free_surface_id' in self.kwargs:
+            log("using surface mass term...")
+            return firedrake.inner(test, trial) * self.ds(self.kwargs['free_surface_id'])
+        else:
+            return firedrake.inner(test, trial) * self.dx
```

Comment:
I think a nicer way to do this is to have FreeSurfaceEquation overload `mass_term()` so you don't have to mess with equation specific things in the base equation.

Reply:
Thanks I've moved the mass term to FreeSurfaceEquation!